state/filesystem.go

```diff
@@ -439,67 +442,92 @@ func (st *State) filesystemAttachments(query bson.D) ([]FilesystemAttachment, er
-	for _, a := range attachments {
-		filesystemTag := a.Filesystem()
-		// When removing the machine, there should only remain
-		// non-persistent storage. This will be implicitly
-		// removed when the machine is removed, so we do not
-		// use removeFilesystemAttachmentOps or removeFilesystemOps,
-		// which track and update related documents.
+	ops := make([]txn.Op, 0, 2*len(machineFilesystems)+len(m.doc.Filesystems))
+	for _, filesystemId := range m.doc.Filesystems {
 		ops = append(ops, txn.Op{
 			C:      filesystemAttachmentsC,
-			Id:     filesystemAttachmentId(machine.Id(), filesystemTag.Id()),
+			Id:     filesystemAttachmentId(m.Id(), filesystemId),
 			Assert: txn.DocExists,
 			Remove: true,
 		})
-		canRemove, err := isFilesystemInherentlyMachineBound(st, filesystemTag)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		if !canRemove {
-			return nil, errors.Errorf("machine has non-machine bound filesystem %v", filesystemTag.Id())
-		}
+	}
+	for _, v := range machineFilesystems {
+		filesystemId := v.Tag().Id()
 		ops = append(ops,
 			txn.Op{
 				C:      filesystemsC,
-				Id:     filesystemTag.Id(),
+				Id:     filesystemId,
 				Assert: txn.DocExists,
 				Remove: true,
 			},
-			removeModelFilesystemRefOp(st, filesystemTag.Id()),
+			removeModelFilesystemRefOp(st, filesystemId),
 		)
 	}
 	return ops, nil
 }
 
-// isFilesystemInherentlyMachineBound reports whether or not the filesystem
-// with the specified tag is inherently bound to the lifetime of the machine,
-// and will be removed along with it, leaving no resources dangling.
-func isFilesystemInherentlyMachineBound(st *State, tag names.FilesystemTag) (bool, error) {
-	// TODO(axw) when we have support for persistent filesystems,
-	// e.g. NFS shares, then we need to check the filesystem info
-	// to decide whether or not to remove.
-	return true, nil
+// isDetachableFilesystemTag reports whether or not the filesystem with the
+// specified tag is detachable.
+func isDetachableFilesystemTag(st *State, tag names.FilesystemTag) (bool, error) {
+	f, err := st.filesystemByTag(tag)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return f.detachable(), nil
+}
+
+// detachable reports whether or not the filesystem is detachable.
+func (f *filesystem) detachable() bool {
+	return f.doc.MachineId == ""
 }
 
-// isFilesystemParamsInherentlyMachineBound reports whether or not the given
-// filesystem params will create a filesystem inherently bound to the lifetime
-// of a machine, and will be removed along with it, leaving no resources dangling.
-func isFilesystemParamsInherentlyMachineBound(st *State, params FilesystemParams) (bool, error) {
-	// TODO(axw) when we have support for persistent filesystems,
-	// e.g. NFS shares, then we need to check the filesystem info
-	// to decide whether or not to remove.
+// isDetachableFilesystemPool reports whether or not the given
+// storage pool will create a filesystem that is not inherently
```

Comment:
"is not inherently a machine" seems clunky
"is not inherently bound to a machine" perhaps?